File: frontend/src/types/entries.ts
/**
 * 가계부 내역 관련 타입 정의
 * 백엔드 models/entries.py와 동기화
 */

export type EntryFrequency = 'once' | 'monthly' | 'weekly';

export interface Entry {
  id: string;
  book_id: string;
  user_id: string;
  entry_date: string; // ISO date format (YYYY-MM-DD)
  description: string;
  amount: number; // 단위: 원
  category: string | null;
  end_date: string | null; // ISO date format (YYYY-MM-DD) - 반복 종료일
  frequency: EntryFrequency; // 반복 주기
  day_of_month: number | null; // 월간 반복 날짜 (1-31)
  day_of_week: number | null; // 주간 반복 요일 (0=일요일, 6=토요일)
  created_at: string; // ISO datetime
  updated_at: string; // ISO datetime
}

export interface EntryCreate {
  entry_date: string; // ISO date format (YYYY-MM-DD)
  description: string; // 1-200자
  amount: number; // 0이 아닌 값
  category?: string | null; // 최대 80자
  end_date?: string | null; // 반복 종료일
  frequency?: EntryFrequency; // 반복 주기 (기본값: 'once')
  day_of_month?: number | null; // 월간 반복 날짜 (1-31)
  day_of_week?: number | null; // 주간 반복 요일 (0-6)
}

export interface EntryUpdate {
  entry_date: string;
  description: string;
  amount: number;
  category?: string | null;
  end_date?: string | null; // 반복 종료일
  frequency?: EntryFrequency; // 반복 주기 (기본값: 'once')
  day_of_month?: number | null; // 월간 반복 날짜 (1-31)
  day_of_week?: number | null; // 주간 반복 요일 (0-6)
}

/**
 * 전개된 내역 (반복 내역의 특정 발생일 인스턴스)
 */
export interface ExpandedEntry extends Entry {
  original_id: string; // 원본 entry ID
  occurrence_date: string; // 실제 발생 날짜 (ISO date)
  is_projected: boolean; // 미래 내역 여부
}

export enum EntryHistoryAction {
  CREATED = 'created',
  UPDATED = 'updated',
  DELETED = 'deleted',
  RESTORED = 'restored',
}

export interface EntryHistoryItem {
  id: string;
  entry_id: string | null;
  book_id: string;
  changed_by: string | null;
  changed_at: string; // ISO datetime
  action_type: EntryHistoryAction;
  snapshot: Record<string, unknown>;
}

export enum EntryTypeFilter {
  INCOME = 'income',
  EXPENSE = 'expense',
}

export interface EntryStatsSummary {
  total_income: number;
  total_expense: number;
  net_amount: number;
}

export interface EntryStatsCategory {
  category: string;
  amount: number;
  ratio: number;
}

export interface EntryStatsTrendPoint {
  period: string; // ISO date (YYYY-MM-01)
  income: number;
  expense: number;
}

export interface EntryStatsTopEntry {
  id: string;
  description: string;
  amount: number;
  entry_date: string;
  category: string | null;
}

export interface EntryStatsResponse {
  summary: EntryStatsSummary;
  category_distribution: EntryStatsCategory[];
  trend: EntryStatsTrendPoint[];
  top_expenses: EntryStatsTopEntry[];
  total_entries: number;
}

export interface EntryBulkImportResultItem {
  index: number;
  success: boolean;
  entry?: Entry;
  error?: string;
}

export interface EntryBulkImportResult {
  total: number;
  success_count: number;
  failure_count: number;
  rows: EntryBulkImportResultItem[];
}

File: backend/app/models/entries.py
"""가계부 내역 및 히스토리 모델."""

from __future__ import annotations

from datetime import date, datetime
from enum import Enum
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


class Entry(BaseModel):
    """가계부 내역 단일 항목."""

    id: UUID
    book_id: UUID
    user_id: UUID
    entry_date: date
    description: str
    amount: int
    category: str | None
    end_date: date | None = None
    frequency: Literal["once", "monthly", "weekly"] = "once"
    day_of_month: int | None = None
    day_of_week: int | None = None
    created_at: datetime
    updated_at: datetime


class EntryCreate(BaseModel):
    """가계부 내역 생성 요청 본문."""

    entry_date: date
    description: str = Field(min_length=1, max_length=200)
    amount: int = Field(description="금액(단위: 원)")
    category: str | None = Field(default=None, max_length=80)
    end_date: date | None = Field(default=None, description="종료일 (반복 내역용)")
    frequency: Literal["once", "monthly", "weekly"] = Field(default="once", description="반복 주기")
    day_of_month: int | None = Field(default=None, ge=1, le=31, description="월간 반복 날짜 (1-31)")
    day_of_week: int | None = Field(
        default=None, ge=0, le=6, description="주간 반복 요일 (0=일요일, 6=토요일)"
    )

    @field_validator("amount")
    @classmethod
    def amount_must_not_be_zero(cls, v: int) -> int:
        """금액은 0이 될 수 없습니다."""
        if v == 0:
            raise ValueError("금액은 0이 될 수 없습니다")
        return v


class EntryType(str, Enum):
    """내역 유형."""

    INCOME = "income"
    EXPENSE = "expense"


class EntryUpdate(BaseModel):
    """가계부 내역 수정 요청 본문."""

    entry_date: date
    description: str = Field(min_length=1, max_length=200)
    amount: int = Field(description="금액(단위: 원)")
    category: str | None = Field(default=None, max_length=80)
    end_date: date | None = Field(default=None, description="종료일 (반복 내역용)")
    frequency: Literal["once", "monthly", "weekly"] = Field(default="once", description="반복 주기")
    day_of_month: int | None = Field(default=None, ge=1, le=31, description="월간 반복 날짜 (1-31)")
    day_of_week: int | None = Field(
        default=None, ge=0, le=6, description="주간 반복 요일 (0=일요일, 6=토요일)"
    )

    @field_validator("amount")
    @classmethod
    def amount_must_not_be_zero(cls, v: int) -> int:
        """금액은 0이 될 수 없습니다."""
        if v == 0:
            raise ValueError("금액은 0이 될 수 없습니다")
        return v


class EntryHistoryAction(str, Enum):
    """히스토리 액션 종류."""

    CREATED = "created"
    UPDATED = "updated"
    DELETED = "deleted"
    RESTORED = "restored"


class EntryHistoryItem(BaseModel):
    """가계부 내역 변경 이력 항목."""

    id: UUID
    entry_id: UUID | None
    book_id: UUID
    changed_by: UUID | None
    changed_at: datetime
    action_type: EntryHistoryAction
    snapshot: dict[str, Any]


class EntryStatsSummary(BaseModel):
    """내역 통계 요약 카드."""

    total_income: int
    total_expense: int
    net_amount: int


class EntryStatsCategory(BaseModel):
    """카테고리별 지표."""

    category: str
    amount: int
    ratio: float


class EntryStatsTrendPoint(BaseModel):
    """월별 추이 포인트."""

    period: date
    income: int
    expense: int


class EntryStatsTopEntry(BaseModel):
    """상위 지출 항목."""

    id: UUID
    description: str
    amount: int
    entry_date: date
    category: str | None


class EntryStats(BaseModel):
    """분석 대시보드 응답."""

    summary: EntryStatsSummary
    category_distribution: list[EntryStatsCategory]
    trend: list[EntryStatsTrendPoint]
    top_expenses: list[EntryStatsTopEntry]
    total_entries: int


class EntryBulkImportRequest(BaseModel):
    """일괄 업로드 요청."""

    rows: list[EntryCreate] = Field(min_length=1, description="업로드할 내역 목록")


class EntryBulkImportResultItem(BaseModel):
    """일괄 업로드 결과 행."""

    index: int
    success: bool
    entry: Entry | None = None
    error: str | None = None


class EntryBulkImportResult(BaseModel):
    """일괄 업로드 결과."""

    total: int
    success_count: int
    failure_count: int
    rows: list[EntryBulkImportResultItem]
